File: CS/WebSite/App_Code/EditableDataItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web.UI;
using DevExpress.Web;
using System.Web.UI.WebControls;

public class EditableDataItem: ITemplate {
    public EditableDataItem() {
    }

    public void InstantiateIn(Control container) {
        GridViewDataItemTemplateContainer dataItemContainer = container as GridViewDataItemTemplateContainer;

        ASPxEdit cellEditor;
        if (dataItemContainer.Column is GridViewDataCheckColumn) {
            cellEditor = new ASPxCheckBox();
        }
        else {
            cellEditor = new ASPxTextBox();
        }

        cellEditor.ID = "edDataItem";
        dataItemContainer.Controls.Add(cellEditor);

        cellEditor.Init += new EventHandler(cellEditor_Init);
        
    }

    void cellEditor_Init(object sender, EventArgs e) {
        ASPxEdit editor = sender as ASPxEdit;
        GridViewDataItemTemplateContainer container = editor.NamingContainer as GridViewDataItemTemplateContainer;

        editor.Width = new Unit(100, UnitType.Percentage);

        ASPxHiddenField hfData = container.Grid.FindTitleTemplateControl("hfData") as ASPxHiddenField;

        string fieldKey = String.Format("{0}_{1}", container.KeyValue, container.Column.FieldName);
        editor.Value = (hfData != null && hfData.Contains(fieldKey)) ? hfData[fieldKey] : DataBinder.Eval(container.DataItem, container.Column.FieldName);

        editor.SetClientSideEventHandler("ValueChanged", String.Format("function(s, e) {{ {0}_hfData.Set('{1}', s.GetValue()); }}", container.Grid.UniqueID, fieldKey)); 
    }


}

File: VB/WebSite/App_Code/EditableDataItem.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web.UI
Imports DevExpress.Web
Imports System.Web.UI.WebControls

Public Class EditableDataItem
	Implements ITemplate
	Public Sub New()
	End Sub

	Public Sub InstantiateIn(ByVal container As Control) Implements ITemplate.InstantiateIn
		Dim dataItemContainer As GridViewDataItemTemplateContainer = TryCast(container, GridViewDataItemTemplateContainer)

		Dim cellEditor As ASPxEdit
		If TypeOf dataItemContainer.Column Is GridViewDataCheckColumn Then
			cellEditor = New ASPxCheckBox()
		Else
			cellEditor = New ASPxTextBox()
		End If

		cellEditor.ID = "edDataItem"
		dataItemContainer.Controls.Add(cellEditor)

		AddHandler cellEditor.Init, AddressOf cellEditor_Init

	End Sub

	Private Sub cellEditor_Init(ByVal sender As Object, ByVal e As EventArgs)
		Dim editor As ASPxEdit = TryCast(sender, ASPxEdit)
		Dim container As GridViewDataItemTemplateContainer = TryCast(editor.NamingContainer, GridViewDataItemTemplateContainer)

		editor.Width = New Unit(100, UnitType.Percentage)

		Dim hfData As ASPxHiddenField = TryCast(container.Grid.FindTitleTemplateControl("hfData"), ASPxHiddenField)

		Dim fieldKey As String = String.Format("{0}_{1}", container.KeyValue, container.Column.FieldName)
		If (hfData IsNot Nothing AndAlso hfData.Contains(fieldKey)) Then
			editor.Value = hfData(fieldKey)
		Else
			editor.Value = DataBinder.Eval(container.DataItem, container.Column.FieldName)
		End If

		editor.SetClientSideEventHandler("ValueChanged", String.Format("function(s, e) {{ {0}_hfData.Set('{1}', s.GetValue()); }}", container.Grid.UniqueID, fieldKey))
	End Sub


End Class